File: app/api/v1/translation.py
"""论文翻译 API (2026-07-20 实装)

端点:
  POST /api/v1/translation/translate
    body: { text: str, target_lang: str }
    response: { translated_text: str, target_lang: str, source_length: int }

复用:
  - app.services.translation_service.translate_text (LLMClient + Redis cache)
  - app.core.security.get_current_user (JWT 鉴权, 与 chat/task 对齐)

错误码:
  - 401: 未登录
  - 422: text 为空 / 超长 / target_lang 不在白名单
  - 500: LLM 服务异常 (兜底由 service 层处理, 此处不抛)
"""
import logging

from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.security import get_current_user
from app.models.member import Member
from app.services.translation_service import translate_text

logger = logging.getLogger("microbubble.api.translation")
router = APIRouter(prefix="/translation", tags=["翻译"])


class TranslateRequest(BaseModel):
    text: str = Field(..., min_length=1, max_length=8000, description="待翻译文本")
    target_lang: str = Field(..., min_length=2, max_length=10, description="目标语言代码 (en/zh/ja/...)")


class TranslateResponse(BaseModel):
    translated_text: str = Field(..., description="翻译结果")
    target_lang: str = Field(..., description="回显目标语言")
    source_length: int = Field(..., description="原文长度 (字符)")


@router.post("/translate", response_model=TranslateResponse)
async def translate(
    body: TranslateRequest,
    db: AsyncSession = Depends(get_db),
    user: Member = Depends(get_current_user),
):
    """翻译文本到目标语言

    - 支持 9 种语言: en / zh / zh-TW / ja / ko / fr / de / es / ru
    - 8000 字符硬上限 (超出返 422)
    - Redis 缓存 24h (同 text+lang 命中)
    - LLM 异常时兜底返原文 (前端不显示空白)
    """
    try:
        translated = await translate_text(body.text, body.target_lang)
    except ValueError as e:
        # 输入校验失败 → 422 (与 chat_service ValueError→422 范式一致)
        from fastapi import HTTPException
        raise HTTPException(status_code=422, detail=str(e))

    return TranslateResponse(
        translated_text=translated,
        target_lang=body.target_lang.strip().lower(),
        source_length=len(body.text),
    )

File: web/src/api/translation.js
/**
 * translation API 客户端 — 2026-07-20 论文段落/全文翻译实装
 *
 * 镜像 web/src/api/chatHistory.js 模式（直接 axios 调用）
 * 1 个端点（来自 app/api/v1/translation.py）：
 *   - POST /translation/translate  翻译 (body: {text, target_lang})
 *
 * 关键纪律：
 * - 失败抛异常 → 调用方 try/except 决定兜底 (默认返原文, 不显示空白)
 * - 422 (text 空/超长/lang 不支持) 抛带 detail 的 Error, 前端 toast 显式提示
 */

import axios from 'axios'

const BASE = '/api/v1'

/**
 * 翻译文本到目标语言
 * @param {Object} params
 * @param {string} params.text - 待翻译文本 (1-8000 字符)
 * @param {string} params.target_lang - 目标语言代码 (en/zh/ja/ko/fr/de/es/ru/zh-TW)
 * @returns {Promise<{translated_text: string, target_lang: string, source_length: number}>}
 * @throws {Error} 422 输入校验失败 / 500 服务异常
 */
export async function translateText({ text, target_lang }) {
  if (!text || !text.trim()) {
    throw new Error('待翻译文本不能为空')
  }
  if (text.length > 8000) {
    throw new Error(`文本超过最大长度 8000 字符 (实际 ${text.length})，请分段翻译`)
  }
  const response = await axios.post(`${BASE}/translation/translate`, {
    text: text.trim(),
    target_lang: target_lang.trim().toLowerCase(),
  })
  return response.data
}
